pvlib/solarposition.py

```diff
@@ -1210,3 +1210,84 @@ def hour_angle(times, longitude, equation_of_time):
     )).total_seconds() / 3600. for t in times])
     timezone = times.tz.utcoffset(times).total_seconds() / 3600.
     return 15. * (hours - 12. - timezone) + longitude + equation_of_time / 4.
+
+
+def _hours(times, hour_angle, longitude, equation_of_time):
+    """helper that converts hour angles in degrees into hours"""
+    tz_info = times.tz  # pytz timezone info
+    tz = tz_info.utcoffset(times).total_seconds() / 3600.
+    return (hour_angle - longitude - equation_of_time / 4.) / 15. + 12. + tz
+
+
+def _times(times, hours):
+    """helper converts hours from an array of floats to localized times"""
+    tz_info = times.tz
+    try:
+        hours = hours.values  # make a new reference
+    except AttributeError:
+        pass
+    try:
+        # OLD pandas, "TypeError: Already tz-aware, use tz_convert to convert."
+        times.tz = None
+    except AttributeError:
+        # NEW pandas, "AttributeError: Cannot directly set timezone. Use
+        # tz_localize() or tz_convert() as appropriate"
+        times = times.tz_localize(None)
+    return pd.DatetimeIndex(
+        times.values.astype('datetime64[D]').astype('datetime64[ns]') +
+        (hours * 3600. * 1.e9).astype('timedelta64[ns]'),
+        tz=tz_info)
+
```

Comment:
Maybe naive, but why not

`return times + pd.to_timedelta(hours * 3600. * 1.e9)`

Reply:
and if that works, then I don't see the value of `_hours`, just do the addition in `sunrise_sunset_transit_analytical`

Reply:
The times have to be reduced to dates first, while preserving their `tz_utcoffset`. There are different ways to do that, my approach here is to set `tz` to `None`, which effectively makes is naive local time, then cast to numpy `datetime64[D]` which is a date.

I'm sure there are a plethora of ways, this one is working and faster than the looping method, so is it okay?

Reply:
Since `times.values.astype('datetime64[D]') extracts the date from `times`, then it is implicit that `times` is localized to the timezone of the desired location.  Otherwise, the algorithm can mix the hour with a different date.  E.g., 

```
ts = pd.DatetimeIndex(['2016-9-20 12:00:00', '2016-9-20 21:00:00']).tz_localize('MST')
tsutc = ts.tz_convert('UTC')
print(ts.values.astype('datetime64[D]'))
['2016-09-20' '2016-09-21']
print(tsutc.values.astype('datetime64[D]'))
['2016-09-20' '2016-09-21']
```

and `_hours` adds the hour value to these dates.

This is also the problem with `get_sun_rise_set_transit`

Reply:
>If I'm correct then the input description of sunrise_sunset_transit_analytical must clarify the restriction on the timezone.

Yes sorry, I should clarify that in the docstring, this function does assume that the times are tz-aware. I mentioned that in [this comment](#issuecomment-424080433).

I guess I thought that this was already a requirement for all pvlib times, not sure why I thought that.

I could relax this requirement though. If the times are naive, and local, then that actually makes the calculations easier, b/c we just use the dates as is, no need to replace `tz` with `None` to convert it to naive if it already is, right?

Reply:
Please disregard the comment I made earlier and then edited while you were replying.  Not a good GitHub workflow on my part.

I think we must require that `times` is localized for this function.  It appears that numpy.datetime64[D] handles the timezone adjustment.  I don't think it is worth the trouble to extend this function to naïve datetimes.

Reply:
it does say it in the docstring already [here](../../../mikofski/pvlib-python/blob/analytical_sunrise_sunset_transit/pvlib/solarposition.py#L1252-L1253)

    times : pandas.DatetimeIndex
        Corresponding timestamps, must be timezone aware.